### PR TITLE
Support aio_pika 8.x

### PR DESCRIPTION
**Pull request by phillipuniverse**
# Description

I am using aio_pika 8.2 and otel currently cannot instrument it:

image

Briefly skimming [the 8.0.0 release notes](https://github.com/mosquito/aio-pika/blob/master/CHANGELOG.md#800) it doesn't seem like the breaking changes would affect instrumentation.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

I would like a bit more information on the right way to do this. Ideally I would like to emit tests for both aio_pika 7.x and 8.x. Tox is not exactly my forte and I'm not sure how to configure this to happen.

That said, since the dependency version is relaxed _theoretically_ the tests will run against 8.x. Need to figure out how to fully verify this (probably related to allowing 7.x and 8.x testing). 

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added - (yes but need an assist)
- [x] Documentation has been updated

### CODE REVIEW

**Comment by srikanthccv** [comment]
>I would like a bit more information on the right way to do this. Ideally I would like to emit tests for both aio_pika 7.x and 8.x. Tox is not exactly my forte and I'm not sure how to configure this to happen.


You will define an env for each version and its corresponding deps. Please take a look at the tox file https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/tox.ini and how it is done for Django.

File: instrumentation/opentelemetry-instrumentation-aio-pika/tests/consts.py
Comment (phillipuniverse) [inline instrumentation/opentelemetry-instrumentation-aio-pika/tests/consts.py:19]: This namespace was a change between 7 and 8

File: instrumentation/opentelemetry-instrumentation-aio-pika/tests/test_publish_decorator.py
Comment (phillipuniverse) [inline instrumentation/opentelemetry-instrumentation-aio-pika/tests/test_publish_decorator.py:113]: This `Exchange` constructor was a change between 7 and 8, and this is also what prompted me to decorate with `skipIf`.

**Review by phillipuniverse — COMMENTED** [review]
A few additional notes:

1. I included tests by copy/pasting and doing a `skipif`. I don't mind refactoring a bit to reduce code duplication if requested. I didn't have a super strong opinion either way but this way was the easiest to implement.
2. I can separate the `black` formatting changes if it's too distracting to review
3. I'm reasonably confident that this encapsulates the changes between aio_pika 7 and 8. I took a look at the classes we were using for the instrumentation and the `Connection` and `Exchange` were the big ones that had changed. I spot-checked the other abstractions we depend on like `AbstractMessage` and those had not changd.

**Comment by phillipuniverse** [comment]
@srikanthccv thanks for the tips! I added separate tox environments for 7 and 8 and iterated a bit on some of the differences in the implementations.

This is now ready for review.

File: instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py
Comment (nozik) [inline instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: Notice that this change shouldn't be necessary - the constructor of `Exchange` doesn't get the connection as its first argument anymore (https://github.com/mosquito/aio-pika/commit/c2ee4be6e4f254e0d36c99f5bcd906cf31fa8233). If you take that into consideration, this change would not be necessary

Comment (phillipuniverse) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: @nozik I'm not sure I'm following, could you be a bit more specific with what you would think this `set_channel` method needs to change to if not what I have here?

Comment (phillipuniverse) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: Ah I see, over in https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1605/files#diff-500514cf8be14e4864fdf49daa4c761c0b1ad0c358cec84b4e26a2c62c1c5dffR57.

That change you made there is incomplete. The change I made here is necessary for aiopika 8.x, the call here fails otherwise:

```python
url = channel.connection.connection.url
```

In your PR, you are still using an aio_pika v7 mock that assumes a property path from `channel.connection.connection.url`. In aio_pika v8, the extra `.connection` is removed, and it's just `channel.connection.url`.

The change occurred actually in the same commit you linked, [these lines](https://github.com/mosquito/aio-pika/commit/c2ee4be6e4f254e0d36c99f5bcd906cf31fa8233#diff-34571b521f08738cafd0fc3ecf1d653376a9f979e8ad04982ff7814a46fcd617L545-R626)

Comment (nozik) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: @phillipuniverse Notice [here](https://github.com/mosquito/aio-pika/commit/c2ee4be6e4f254e0d36c99f5bcd906cf31fa8233) the changes to `exchange.py` - the `Exchange` class no longer receives `AbstractConnection` as its first constructor argument. However, in `test_get_publish_span` we're still sending wrongfully sending it as the first argument. If you just check the condition of the `aio-pika` version and adapt the creation of the `Exchange` (simply remove the first ctor argument for version >= 8.0.0), all the other code changes will become redundant. I hope this helps.

Comment (phillipuniverse) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: There are 2 changes that needed to happen to support aio_pika 8:

1. The change you are describing to the constructor in the test as aio_pika 8 changed the signature
2. The change in this method to deal with the changed shape of AbstractConnection

There aren’t good integration tests for aio_pika yet in this repo but I observed the problem in this method in my integration environment after patching the instrumentation version check for to indiscriminately apply to aio_pika 8

File: instrumentation/opentelemetry-instrumentation-aio-pika/tests/consts.py
Comment (phillipuniverse) [reply instrumentation/opentelemetry-instrumentation-aio-pika/tests/consts.py:19]: @nozik this is the namespace change I was describing above

File: instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py
Comment (nozik) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: 👍

Comment (akalex) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: Hi there,
I'm also looking for support of aio_pika 8 so I wonder, when optimistically this PR could be merged and released? Thank you in advance.

**Comment by srikanthccv** [comment]
Ping @ofek1weiss as component owner and the original author.

File: instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py
Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-aio-pika/src/opentelemetry/instrumentation/aio_pika/span_builder.py:52]: This instrumentation is supported and maintained by community members. Maintainers don't have any expertise in this instrumentation to judge the changes. The approval from the [component owner](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/.github/component_owners.yml) gets it merged immediately and will be part of the next scheduled release.

**Comment by srikanthccv** [comment]
Another friendly ping

**Comment by ofek1weiss** [comment]
Hey, sorry I missed the first ping
It looks like one of the checks failed :(
Once that is fixed everything looks good to me 👍

File: CHANGELOG.md
Comment (srikanthccv) [inline CHANGELOG.md:26]:
```suggestion
```

**Review by srikanthccv — APPROVED** [review]

**Comment by srikanthccv** [comment]
Merging this as it got LGTM from the codeowners.

**Pull request merged**
